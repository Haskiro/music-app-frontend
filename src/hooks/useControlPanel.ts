import { useState } from "react";
import { useAppDispatch } from "@store/hooks";

interface IUseControlPanelResult {
	searchValue: string;
	sortByAlphabet: boolean;
	onSearch: (searchValue: string) => void;
	onSort: () => void;
	onUpdate: () => void;
}

const useControlPanel = (fetchItems: any): IUseControlPanelResult => {
	const [searchValue, setSearchValue] = useState<string>("");
	const [sortByAlphabet, setSortByAlphabet] = useState<boolean>(false);
	const dispatch = useAppDispatch();

	const onSearch = (searchValue: string) => {
		setSearchValue(searchValue.toLowerCase());
	};
	const onSort = () => setSortByAlphabet((prevState) => !prevState);
	const onUpdate = () => dispatch(fetchItems());
	return { searchValue, sortByAlphabet, onSearch, onSort, onUpdate };
};

export default useControlPanel;
