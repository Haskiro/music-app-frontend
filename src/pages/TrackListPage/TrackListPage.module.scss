.heading {
	text-align: center;
	font-size: 1.4rem;
	line-height: 1.2;
	font-weight: 700;
	margin-bottom: 20px;
}
.control {
	&Panel {
		display: flex;
		margin-bottom: 20px;
		justify-content: center;
		gap: 10px;
		@media (max-width: 768px) {
			flex-direction: column;
			align-items: center;
		}
	}
	&Buttons {
		display: flex;
		gap: 10px;
		&Item {
			background: none;
			border: 1px solid lightgray;
			cursor: pointer;
			border-radius: 6px;
			padding: 10px 20px;
			cursor: pointer;
			font-size: 1rem;
			white-space: nowrap;
			&:hover {
				color: cyan;
			}
			&Checked {
				color: cyan;
			}
		}
	}
	&Input {
	}
}

.result {
	text-align: center;
	margin-top: 20px;
	font-size: 1.2rem;
}
