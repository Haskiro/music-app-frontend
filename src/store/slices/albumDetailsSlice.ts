import { IAlbum } from "@interfaces/album.interface";
import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import axios from "axios";

export interface albumDetailsState {
	album: IAlbum | null;
	status: "idle" | "loading" | "succeeded" | "failed";
}

const initialState: albumDetailsState = {
	album: null,
	status: "idle",
};

export const fetchAlbumById = createAsyncThunk(
	"albumDetails/fetchAlbumById",
	async (id: string) => {
		const response = await axios.get(
			`${process.env.REACT_APP_API}/albums/${id}`
		);
		return response.data;
	}
);

export const albumDetailsSlice = createSlice({
	name: "albumDetails",
	initialState,
	reducers: {},
	extraReducers(builder) {
		builder
			.addCase(fetchAlbumById.pending, (state) => {
				state.status = "loading";
			})
			.addCase(
				fetchAlbumById.fulfilled,
				(state, action: PayloadAction<IAlbum>) => {
					state.status = "succeeded";
					state.album = action.payload;
				}
			)
			.addCase(fetchAlbumById.rejected, (state) => {
				state.status = "failed";
			});
	},
});

// Action creators are generated for each case reducer function
export const {} = albumDetailsSlice.actions;

export default albumDetailsSlice.reducer;
