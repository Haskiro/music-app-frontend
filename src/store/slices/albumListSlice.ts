import { IAlbum } from "@interfaces/album.interface";
import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import axios from "axios";

export interface albumListState {
	albumList: IAlbum[] | null;
	status: "idle" | "loading" | "succeeded" | "failed";
}

const initialState: albumListState = {
	albumList: null,
	status: "idle",
};

export const fetchAlbums = createAsyncThunk(
	"albumList/fetchAlbums",
	async () => {
		const response = await axios.get(`${process.env.REACT_APP_API}/albums`);
		return response.data;
	}
);

export const albumListSlice = createSlice({
	name: "artistList",
	initialState,
	reducers: {},
	extraReducers(builder) {
		builder
			.addCase(fetchAlbums.pending, (state) => {
				state.status = "loading";
			})
			.addCase(
				fetchAlbums.fulfilled,
				(state, action: PayloadAction<IAlbum[]>) => {
					state.status = "succeeded";
					state.albumList = action.payload;
				}
			)
			.addCase(fetchAlbums.rejected, (state) => {
				state.status = "failed";
			});
	},
});

// Action creators are generated for each case reducer function
export const {} = albumListSlice.actions;

export default albumListSlice.reducer;
