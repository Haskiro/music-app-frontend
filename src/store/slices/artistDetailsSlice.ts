import { IArtist } from "@interfaces/artist.interface";
import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import axios from "axios";

export interface artistDetailsState {
	artist: IArtist | null;
	status: "idle" | "loading" | "succeeded" | "failed";
}

const initialState: artistDetailsState = {
	artist: null,
	status: "idle",
};

export const fetchArtistById = createAsyncThunk(
	"artistDetails/fetchArtist",
	async (id: string) => {
		const response = await axios.get(
			`${process.env.REACT_APP_API}/artists/${id}`
		);
		return response.data;
	}
);

export const artistDetailsSlice = createSlice({
	name: "artistDetails",
	initialState,
	reducers: {},
	extraReducers(builder) {
		builder
			.addCase(fetchArtistById.pending, (state) => {
				state.status = "loading";
			})
			.addCase(
				fetchArtistById.fulfilled,
				(state, action: PayloadAction<IArtist>) => {
					state.status = "succeeded";
					state.artist = action.payload;
				}
			)
			.addCase(fetchArtistById.rejected, (state) => {
				state.status = "failed";
			});
	},
});

// Action creators are generated for each case reducer function
export const {} = artistDetailsSlice.actions;

export default artistDetailsSlice.reducer;
