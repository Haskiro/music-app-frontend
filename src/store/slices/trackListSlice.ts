import { ITrack } from "@interfaces/track.interface";
import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";
import axios from "axios";

export interface trackListState {
	trackList: ITrack[] | null;
	status: "idle" | "loading" | "succeeded" | "failed";
}

const initialState: trackListState = {
	trackList: null,
	status: "idle",
};

export const fetchTracks = createAsyncThunk(
	"trackList/fetchTracks",
	async () => {
		const response = await axios.get(`${process.env.REACT_APP_API}/tracks`);
		return response.data;
	}
);

export const trackListSlice = createSlice({
	name: "trackList",
	initialState,
	reducers: {},
	extraReducers(builder) {
		builder
			.addCase(fetchTracks.pending, (state) => {
				state.status = "loading";
			})
			.addCase(
				fetchTracks.fulfilled,
				(state, action: PayloadAction<ITrack[]>) => {
					state.status = "succeeded";
					state.trackList = action.payload;
				}
			)
			.addCase(fetchTracks.rejected, (state) => {
				state.status = "failed";
			});
	},
});

// Action creators are generated for each case reducer function
export const {} = trackListSlice.actions;

export default trackListSlice.reducer;
