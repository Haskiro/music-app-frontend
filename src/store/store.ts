import { configureStore, combineReducers } from "@reduxjs/toolkit";
import {
	persistStore,
	persistReducer,
	FLUSH,
	REHYDRATE,
	PAUSE,
	PERSIST,
	PURGE,
	REGISTER,
} from "redux-persist";
import storage from "redux-persist/lib/storage";
import userReducer from "./slices/userSlice";

const persistConfig = {
	key: "user",
	version: 1,
	storage,
	whitelist: ["user"],
};

const userPersistConfig = {
	key: "accessToken",
	storage,
	whitelist: ["accessToken"],
};

const rootReducer = combineReducers({
	user: persistReducer(userPersistConfig, userReducer),
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
	reducer: persistedReducer,
	middleware: (getDefaultMiddleware) =>
		getDefaultMiddleware({
			serializableCheck: {
				ignoredActions: [
					FLUSH,
					REHYDRATE,
					PAUSE,
					PERSIST,
					PURGE,
					REGISTER,
				],
			},
		}),
});

export let persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
