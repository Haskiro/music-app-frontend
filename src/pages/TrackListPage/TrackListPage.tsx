import { FC, useEffect, useMemo } from "react";
import TrackList from "@components/TrackList";
import useControlPanel from "@hooks/useControlPanel";
import { useAppDispatch, useAppSelector } from "@store/hooks";
import { fetchTracks } from "@store/slices/trackListSlice";
import filterItemList from "@utils/filterItemList";
import { Input } from "antd";
import cn from "classnames";
import styles from "./TrackListPage.module.scss";

const TrackListPage: FC = () => {
	const { Search } = Input;
	const dispatch = useAppDispatch();
	const tracks = useAppSelector((state) => state.trackList.trackList);
	const status = useAppSelector((state) => state.trackList.status);
	const { searchValue, sortByAlphabet, onSearch, onSort, onUpdate } =
		useControlPanel(fetchTracks);

	useEffect(() => {
		dispatch(fetchTracks());
	}, []);

	const filteredTracks = useMemo(() => {
		let filteredTracks = tracks.slice();

		return filterItemList(
			filteredTracks,
			searchValue,
			sortByAlphabet,
			"title"
		);
	}, [tracks, sortByAlphabet, searchValue]);

	return (
		<>
			<h1 className={styles.heading}>Треки</h1>
			<div className={styles.controlPanel}>
				<Search
					className={styles.controlInput}
					placeholder="Введите название трека"
					allowClear
					enterButton="Поиск"
					size="large"
					onSearch={onSearch}
				/>
				<div className={styles.controlButtons}>
					<button
						className={cn(styles.controlButtonsItem, {
							[styles.controlButtonsItemChecked]: sortByAlphabet,
						})}
						onClick={onSort}
					>
						По Алфавиту
					</button>
					<button
						className={styles.controlButtonsItem}
						onClick={onUpdate}
					>
						Обновить
					</button>
				</div>
			</div>
			{filteredTracks.length === 0 ? (
				<p className={styles.result}>Ничего не найдено</p>
			) : (
				<TrackList trackList={filteredTracks} status={status} />
			)}
		</>
	);
};

export default TrackListPage;
