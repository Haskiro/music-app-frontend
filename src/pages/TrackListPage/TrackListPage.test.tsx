import { render } from "@testing-library/react";
import * as reduxHooks from "../../store/hooks";
import * as actions from "../../store/slices/trackListSlice";
import TrackListPage from "./TrackListPage";

jest.mock("../../store/hooks");

/** 
	const mockedUseAppSelector = jest.fn(() => []);
	jest.mock("../../store/hooks", () => ({
		useAppSelector: () => mockedUseAppSelector(),
	}));
*/

/** 
 	const mockedUseAppSelector = useAppSelector as jest.Mocked<
		typeof useAppSelector
	>; 
*/

const mockedUseAppSelector = jest.spyOn(reduxHooks, "useAppSelector");
const mockedUseAppDispatch = jest.spyOn(reduxHooks, "useAppDispatch");

describe("Track List Page", () => {
	it("should create Track list page with empty tracks", () => {
		// Only js files, no typescript
		// mockedUseAppSelector.mockReturnValue([]);

		mockedUseAppSelector.mockReturnValue([]);
		mockedUseAppDispatch.mockReturnValue(jest.fn());
		const component = render(<TrackListPage />);

		expect(component).toMatchSnapshot();
	});
});
