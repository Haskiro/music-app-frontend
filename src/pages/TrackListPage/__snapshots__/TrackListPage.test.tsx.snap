// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track List Page should create Track list page with empty tracks 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        class="heading"
      >
        Треки
      </h1>
      <div
        class="controlPanel"
      >
        <span
          class="ant-input-group-wrapper ant-input-search ant-input-search-large ant-input-search-with-button controlInput ant-input-group-wrapper-lg css-dev-only-do-not-override-sk7ap8"
        >
          <span
            class="ant-input-wrapper ant-input-group css-dev-only-do-not-override-sk7ap8"
          >
            <span
              class="ant-input-affix-wrapper ant-input-affix-wrapper-lg css-dev-only-do-not-override-sk7ap8"
            >
              <input
                class="ant-input ant-input-lg css-dev-only-do-not-override-sk7ap8"
                placeholder="Поиск по странице"
                type="text"
                value=""
              />
              <span
                class="ant-input-suffix"
              >
                <span
                  class="ant-input-clear-icon ant-input-clear-icon-hidden"
                  role="button"
                  tabindex="-1"
                >
                  <span
                    aria-label="close-circle"
                    class="anticon anticon-close-circle"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="close-circle"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                      />
                    </svg>
                  </span>
                </span>
              </span>
            </span>
            <span
              class="ant-input-group-addon"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-sk7ap8 ant-btn-primary ant-btn-lg ant-input-search-button"
                type="button"
              >
                <span>
                  Поиск
                </span>
              </button>
            </span>
          </span>
        </span>
        <div
          class="controlButtons"
        >
          <button
            class="controlButtonsItem"
            data-testid="sortButton"
          >
            По Алфавиту
          </button>
          <button
            class="controlButtonsItem"
          >
            Обновить
          </button>
        </div>
      </div>
      <p
        class="result"
      >
        Ничего не найдено
      </p>
    </div>
  </body>,
  "container": <div>
    <h1
      class="heading"
    >
      Треки
    </h1>
    <div
      class="controlPanel"
    >
      <span
        class="ant-input-group-wrapper ant-input-search ant-input-search-large ant-input-search-with-button controlInput ant-input-group-wrapper-lg css-dev-only-do-not-override-sk7ap8"
      >
        <span
          class="ant-input-wrapper ant-input-group css-dev-only-do-not-override-sk7ap8"
        >
          <span
            class="ant-input-affix-wrapper ant-input-affix-wrapper-lg css-dev-only-do-not-override-sk7ap8"
          >
            <input
              class="ant-input ant-input-lg css-dev-only-do-not-override-sk7ap8"
              placeholder="Поиск по странице"
              type="text"
              value=""
            />
            <span
              class="ant-input-suffix"
            >
              <span
                class="ant-input-clear-icon ant-input-clear-icon-hidden"
                role="button"
                tabindex="-1"
              >
                <span
                  aria-label="close-circle"
                  class="anticon anticon-close-circle"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="close-circle"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                    />
                  </svg>
                </span>
              </span>
            </span>
          </span>
          <span
            class="ant-input-group-addon"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-sk7ap8 ant-btn-primary ant-btn-lg ant-input-search-button"
              type="button"
            >
              <span>
                Поиск
              </span>
            </button>
          </span>
        </span>
      </span>
      <div
        class="controlButtons"
      >
        <button
          class="controlButtonsItem"
          data-testid="sortButton"
        >
          По Алфавиту
        </button>
        <button
          class="controlButtonsItem"
        >
          Обновить
        </button>
      </div>
    </div>
    <p
      class="result"
    >
      Ничего не найдено
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
